app/api/auth/[...nextauth]/route.ts

```diff
@@ -0,0 +1,107 @@
+import NextAuth, { type DefaultUser, type NextAuthOptions } from 'next-auth';
+import GithubProvider from 'next-auth/providers/github';
+import GoogleProvider from 'next-auth/providers/google';
+import CredentialsProvider from 'next-auth/providers/credentials';
+
+import User, { IUser } from '@/models/User';
+import { connectToMongoDB } from '@/lib/database/mongodb';
+
+// ----------------------------------------------------------------
+
+const authOptions: NextAuthOptions = {
+  providers: [
+    GithubProvider({
+      clientId: process.env.GITHUB_ID ?? '',
+      clientSecret: process.env.GITHUB_SECRET ?? '',
+    }),
+    GoogleProvider({
+      clientId: process.env.GOOGLE_ID ?? '',
+      clientSecret: process.env.GOOGLE_SECRET ?? '',
+    }),
+    CredentialsProvider({
+      name: 'signin',
+      id: 'signin',
+      credentials: {
+        email: {
+          label: 'Email',
+          type: 'text',
+          placeholder: 'Enter your email address',
+        },
+        password: { label: 'Password', type: 'password' },
+      },
+      async authorize(credentials, req) {
+        console.log('credentials', credentials);
+
+        let user: IUser | null = null;
+
+        try {
+          await connectToMongoDB();
+
+          // ? Should the password be encrypted when
+          user = await User.findOne({
+            email: credentials?.email,
+            password: credentials?.password,
+          });
```

Comment:
Of course, we can't store plain text passwords in the database, all of them must be encrypted (hashed). A good option is using the `bcrypt` library to hash the passwords.